dp/cloud/python/magma/configuration_controller/crl_validator/certificate.py

```diff
@@ -10,46 +10,56 @@
 See the License for the specific language governing permissions and
 limitations under the License.
 """
+import logging
 import socket
 import ssl
-from typing import List
+from typing import List, Optional
 
 import requests
 from cryptography import x509
-from cryptography.hazmat.backends import default_backend
-from requests.exceptions import SSLError
+from requests.exceptions import RequestException, SSLError
 
-_x509_backend = default_backend()
+logger = logging.getLogger(__name__)
 
 
-def get_certificate(
-        hostname: str,
-        port: int = 443,
-) -> x509.Certificate:
-    """ Retrieve certificate of given host.
+def get_certificate(hostname: str, port: int = 443) -> Optional[x509.Certificate]:
+    """ Retrieve SSL certificate of given host.
 
     Args:
         hostname: host from which certificate will be retrieved
         port: SSL port of the host
 
+    Raises:
+        ConnectionError: if there was an error while trying to get certificate from host
+        ValueError: if data received from host is not valid DER certificate
+
     Returns:
         x509.Certificate
     """
     context = ssl.SSLContext(protocol=ssl.PROTOCOL_SSLv23)
     context.minimum_version = ssl.TLSVersion.TLSv1_2
 
-    with socket.create_connection((hostname, port)) as sock:
-        sock = context.wrap_socket(sock, server_hostname=hostname)
-        binary_certificate = sock.getpeercert(binary_form=True)
-        sock.shutdown(socket.SHUT_RDWR)
+    try:
+        with socket.create_connection((hostname, port)) as sock:
+            sock = context.wrap_socket(sock, server_hostname=hostname)
+            binary_certificate = sock.getpeercert(binary_form=True)
+            sock.shutdown(socket.SHUT_RDWR)
+    except socket.error as e:
+        msg = f'Unable to get SSL certificate for {hostname}: {e}'
+        logger.warning(msg)
```

Comment:
we are logging this message multiple times (here and in `get_certificate_crls`, can we log only there (or maybe in top level not to have logs cluttered)

Reply:
Moved this logging from `get_certificate` to `update_certificates`.   
Error in `get_certificate_crls` is about being unable to get CRL, not the cert itself.